### PR TITLE
Netlist construction from schematic

### LINKED ISSUE #118: Construct netlist from IR

### PR DESCRIPTION
**Pull request by mtrberzi**
Fixes #118. There are couple of TODOs here that I would like some comments/suggestions on before this is merged, but thanks to the power of sharing, we now have netlist building in about 100 lines of code.

### CODE REVIEW

**Comment by mtrberzi** [comment]
Would love to get this code-reviewed and merged once networks is done. I'm going to start Design Rule Checks independently on the assumption that this gets merged more or less as is (i.e. feature-branch off of this and then rebase onto master once this gets merged).

**Comment by mtrberzi** [comment]
Just to let everyone know, I have a branch for DRC ready to go _and_ a branch for code generation nearly ready -- and they both depend on everything in this PR.

File: src/main/java/org/manifold/compiler/back/digital/Netlist.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    if (!connection.getType().equals(digitalWireType)) {
+      throw new TypeMismatchException(digitalWireType, connection.getType());
+    }
+  }
+  
+  private void verify_PortIsDigitalIO(PortValue port) 
+      throws TypeMismatchException {
+    if (!(port.getType().equals(digitalInType) 
+        || port.getType().equals(digitalOutType))) {
+      // TODO this exception is inaccurate because there are two valid types
```
Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/back/digital/Netlist.java]: Would it be possible to resolve this before committing?

File: src/main/java/org/manifold/compiler/back/digital/Netlist.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+          Net newNet = new Net(netName);
+          nets.put(netName, newNet);
+          connectToNet(portFrom, newNet);
+          connectToNet(portTo, newNet);
+        }
+      }
+    }
+  }
+  
+  private void verify_ConnectionIsDigitalWire(ConnectionValue connection)
```
Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/back/digital/Netlist.java]: All methods, other than tests, should be written in camelCase.

File: src/main/java/org/manifold/compiler/back/digital/Net.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  public Net(String name) {
+    this.name = name;
+  }
+  
+  private Set<PortValue> connectedPorts = new HashSet<>();  
+  public Set<PortValue> getConnectedPorts() {
+    return ImmutableSet.copyOf(connectedPorts);
+  }
+  public void addPort(PortValue port){
+    // TODO(murphy) what happens if the port is already connected to this net?
```
Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/back/digital/Net.java]: Perhaps you should throw an exception in this case.

File: src/main/java/org/manifold/compiler/back/digital/Netlist.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  private Map<String, Net> nets = new HashMap<>();
+  public Map<String, Net> getNets() {
+    return ImmutableMap.copyOf(nets);
+  }
+  private Map<PortValue, Net> connectedNet = new HashMap<>();
+  public Net getConnectedNet(PortValue port) {
+    if (connectedNet.containsKey(port)) {
+      return connectedNet.get(port);
+    } else {
+      // TODO(murphy) throw an exception?
```
Comment (lucaswoj) [inline src/main/java/org/manifold/compiler/back/digital/Netlist.java]: Please do. Better to throw the exception and remove it later if it becomes annoying/problematic.

**Comment by lucaswoj** [comment]
This is great! Thanks for keeping this nice and concise. 

I think you should throw exceptions in all the places where you're considering it. Better to throw than to not know! (That rhymes so its probably true).

**Comment by mtrberzi** [comment]
Thanks! Good call for exceptions, I will insert those and fix these other nits, and then I think this will be merged without issue.

**Pull request merged**